scripts/install-mender.sh

```diff
@@ -152,21 +195,82 @@ add_repo() {
     echo "$repo" > /etc/apt/sources.list.d/mender.list
 }
 
-do_install() {
+do_install_open() {
+    # Filter out commercial components
+    local selected_components_open=""
+    for c in $SELECTED_COMPONENTS
+    do
+        if ! echo "$COMMERCIAL_COMPONENTS" | egrep -q "(^| )$c( |\$)"; then
+            selected_components_open="$selected_components_open $c"
+        fi
+    done
+
+    # Return if no open source components selected
+    if [ -z "$selected_components_open" ]; then
+        return
+    fi
+
+    echo "  Installing open source components from APT repository"
+
     apt-get update
     apt-get install -y \
        -o Dpkg::Options::="--force-confdef" \
        -o Dpkg::Options::="--force-confold" \
-       $SELECTED_COMPONENTS
+       $selected_components_open
+
+    echo "  Success! Please run \`mender setup\` to configure the client."
+}
+
+do_install_commercial() {
+    # Filter commercial components
+    local selected_components_commercial=""
+    for c in $SELECTED_COMPONENTS
+    do
+        if echo "$COMMERCIAL_COMPONENTS" | egrep -q "(^| )$c( |\$)"; then
+            selected_components_commercial="$selected_components_commercial $c"
+        fi
+    done
+
+    # Return if no commercial components selected
+    if [ -z "$selected_components_commercial" ]; then
+        return
+    fi
+
+    echo "  Installing commercial components from $MENDER_COMMERCIAL_DOWNLOAD_URL"
+
+    # Translate Debian "channel" into Mender version
+    if [ "$CHANNEL" = "experimental" ]; then
+        version="master"
+    else
+        version="latest"
+    fi
+
+    # Download deb packages
+    for c in $selected_components_commercial
+    do
+        url="$MENDER_COMMERCIAL_DOWNLOAD_URL$(printf ${COMMERCIAL_COMP_TO_URL_PATH_F[$c]} $version $version)"
+        curl -fLsS -H "Authorization: Bearer $JWT_TOKEN" -O $url
+    done
+
+    # Install all of them at once and fallback to install missing dependencies
+    local deb_packages_glob=$(echo $selected_components_commercial | sed -e 's/ /*.deb /g; s/$/*.deb/')
+    dpkg --install $deb_packages_glob || apt-get -f install
+
+    # Check individually each package
+    for c in $selected_components_commercial
+    do
```

Comment:
I wonder why we do not use the for loop and do `dpkg -i  "$c"*deb` and then `--status`? do commercial components depend on each other?

Reply:
> I wonder why we do not use the for loop and do `dpkg -i "$c"*deb` and then `--status`? do commercial components depend on each other? 

That is what I am considering, that commercial components _could_ depend on each other in the future, and then the order on which the user specifies them will make the dpkg call pass or fail

> technically you can drop --status, if -i was successful, right?

Not really. Even if they were independent on each other, some packages might have extra dependencies (for example `mender-monitor` depends on `lmdb-utils`) and hence we need to catch and error and try amending dependencies with `|| apt-get -f install`. And then, the only correct way to check that the pacakge was insalled is with `--status`.

I couldn't find a way to specify up front to dpkg to install the dependencies. Feel free to read its `man` page :stuck_out_tongue: